{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "17f7835f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import numpy as py\n",
    "class FastAreader :\n",
    "    \n",
    "    def __init__ (self, fname=''):\n",
    "        '''contructor: saves attribute fname '''\n",
    "        \n",
    "        self.fname = fname\n",
    "            \n",
    "    def doOpen (self):\n",
    "        if self.fname is '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    " \n",
    "    def readFasta (self):\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        \n",
    "        with self.doOpen() as fileH:\n",
    "            header = ''\n",
    "            sequence = ''\n",
    " \n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>') :\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    " \n",
    "            for line in fileH:\n",
    "                if line.startswith ('>'):\n",
    "                    yield header,sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else :\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "        yield header,sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "b61f53e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "########################################################################\n",
    "# File:program.py\n",
    "#  executable: program.py\n",
    "# Purpose:scaffold file for python executables\n",
    "#   stderr: errors and status\n",
    "#   stdout:\n",
    "#          \n",
    "# Author: David Bernick\n",
    "# History:      dlb 08/20/2011 Created\n",
    "#               dlb 08/26/2021 minor updates\n",
    "########################################################################\n",
    "\n",
    "########################################################################\n",
    "# CommandLine\n",
    "########################################################################\n",
    "class CommandLine() :\n",
    "    '''\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse,\n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination exception Usage.\n",
    "\n",
    "    attributes:\n",
    "    all arguments received from the commandline using .add_argument will be\n",
    "    available within the .args attribute of an object instantiated from CommandLine.\n",
    "    For example, if thisCommandLine is an object of the class, and requiredbool was\n",
    "    set as an option using add_argument, then thisCommandLine.args.requiredbool will\n",
    "    name that option.\n",
    " \n",
    "    '''\n",
    "    \n",
    "    def __init__(self, inOpts=None) :\n",
    "        '''\n",
    "        CommandLine constructor.\n",
    "        \n",
    "        Implement a parser to interpret the command line argv string using argparse.\n",
    "        '''\n",
    "        \n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(description = 'Program prolog - a brief description of what this thing does', \n",
    "                                             epilog = 'Program epilog - some other stuff you feel compelled to say', \n",
    "                                             add_help = True, #default is True \n",
    "                                             prefix_chars = '-', \n",
    "                                             usage = '%(prog)s [options] -option1[default] <input >output' \n",
    "                                             )\n",
    "        self.parser.add_argument('-i', '--iterations', type=int, action = 'store')\n",
    "        self.parser.add_argument('-p', '--psuedocount', type=int, action = 'store')\n",
    "        self.parser.add_argument('-k', '--length', type=int, action = 'store')\n",
    "        if inOpts is None :\n",
    "            self.args = self.parser.parse_args()\n",
    "        else :\n",
    "            self.args = self.parser.parse_args(inOpts)\n",
    "  \n",
    "\n",
    "class Usage(Exception):\n",
    "    '''\n",
    "    Signal a Usage error, evoking a usage statement and eventual exit when raised.\n",
    "    '''\n",
    "    def __init__(self, msg):\n",
    "        self.msg = msg\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 736,
   "id": "3d9357df",
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# The Bioinformatics Class\n",
    "# Here is where most of your solution will end up.\n",
    "#\n",
    "########################################################################\n",
    "class NullModel:\n",
    "    '''\n",
    "    Build the thing that needs doing.\n",
    "    '''\n",
    "    def __init__(self):\n",
    "        self.nucleotideProb = {}\n",
    "        \n",
    "    \n",
    "    def prob(self, seqs):\n",
    "        '''\n",
    "        '''\n",
    "        total = ''.join(seqs)\n",
    "        lenSeqData = len(total)\n",
    "        nucPr = {}\n",
    "        for nucleotide in total:\n",
    "            if nucleotide not in nucPr.keys():\n",
    "                nucPr[nucleotide] = 1\n",
    "            else:\n",
    "                nucPr[nucleotide] += 1\n",
    "        for nucleotide, count in nucPr.items():\n",
    "            nucPr[nucleotide] = count/lenSeqData\n",
    "            \n",
    "        self.nucleotideProb = nucPr\n",
    "        return self.nucleotideProb\n",
    "                \n",
    "########################################################################\n",
    "# The Bioinformatics Class\n",
    "# Here is where most of your solution will end up.\n",
    "#\n",
    "########################################################################\n",
    "class RandomMotif:\n",
    "    '''\n",
    "    Build the thing that needs doing.\n",
    "    '''\n",
    "    def __init__(self):\n",
    "        self._motif = ''\n",
    "\n",
    "    def motif(self, seq, k):\n",
    "        index = np.random.randint(0, len(seq) - k)\n",
    "        window = seq[index : index+k]\n",
    "        \n",
    "        self._motif = window\n",
    "        return self._motif\n",
    "        \n",
    "        \n",
    "########################################################################\n",
    "# The Bioinformatics Class\n",
    "# Here is where most of your solution will end up.\n",
    "#\n",
    "########################################################################\n",
    "class Profile:\n",
    "    '''\n",
    "    Build the thing that needs doing.\n",
    "    '''\n",
    "    def __init__(self):\n",
    "        self.profile = []\n",
    "\n",
    "    \n",
    "    def countMatrix (self, seqs, motiflist, k, p):\n",
    "        matrix = []\n",
    "        for i in range(k):\n",
    "            matrix.append({'A': p, 'T': p, 'C': p, 'G': p})\n",
    "        for index, motif in enumerate(motiflist):\n",
    "            for i, nucleotide in enumerate(motif):\n",
    "                matrix[i][nucleotide] += 1\n",
    "        for row in matrix:\n",
    "            for nucleotide in row:\n",
    "                row[nucleotide] = row[nucleotide] / (len(seqs) + (p*4))\n",
    "        self.profile = matrix\n",
    "        \n",
    "        return self.profile\n",
    "            \n",
    "\n",
    "            \n",
    "########################################################################\n",
    "# The Bioinformatics Class\n",
    "# Here is where most of your solution will end up.\n",
    "#\n",
    "########################################################################\n",
    "class MotifDict:\n",
    "    '''\n",
    "    '''\n",
    "    def __init__(self):\n",
    "        self.probs = []\n",
    "        self.newMotifs = []\n",
    "        \n",
    "    def probability (self, seqs, profileMatrix, k):\n",
    "        prob = []\n",
    "        newMotifs = []\n",
    "        #print(6, newMotifs)\n",
    "        for seq in seqs:\n",
    "            #print (7, seq)\n",
    "            dictionary = {}\n",
    "            windowMax = ('', 0)\n",
    "            for i in range(len(seq)-k+1):\n",
    "                window = [seq[i:i+k]][0]\n",
    "                dictionary[window] = 1\n",
    "                for j, nucleotide in enumerate(window):\n",
    "                    dictionary[window] = (dictionary[window] * \n",
    "                    profileMatrix[j][nucleotide])\n",
    "                #print (8, window, dictionary[window])\n",
    "            \n",
    "                if dictionary[window] > windowMax[1]:\n",
    "                    windowMax = (window, dictionary[window])\n",
    "            #print(9, windowMax)\n",
    "            newMotifs.append(windowMax[0])\n",
    "            #print(10, newMotifs)\n",
    "            prob.append(dictionary)\n",
    "            #print(dictionary)\n",
    "            self.newMotifs =newMotifs\n",
    "        \n",
    "        return self.newMotifs\n",
    "\n",
    "########################################################################\n",
    "# The Bioinformatics Class\n",
    "# Here is where most of your solution will end up.\n",
    "#\n",
    "########################################################################\n",
    "class MotifFinder:\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.consensus = ''\n",
    "        self.score = 0\n",
    "        \n",
    "    def motifSearch(self, motiflist, seqs, k, p):\n",
    "            \n",
    "                \n",
    "        profileMatrix = Profile().countMatrix(seqs, motiflist, k, p)\n",
    "        #print(profileMatrix)   \n",
    "        consensus = ''\n",
    "        RelativeEntropy = 0\n",
    "        Q = NullModel().prob(seqs)\n",
    "        for row in profileMatrix:\n",
    "            consensus += max(row, key=row.get)\n",
    "            #print (consensus)\n",
    "            for nucleotide in row:\n",
    "                RelativeEntropy -= (row[nucleotide] * np.log2(row[nucleotide]) * (row[nucleotide] / Q[nucleotide]))   \n",
    "        return (consensus, RelativeEntropy)\n",
    "    \n",
    "        \n",
    "    def iterate(self, seqs, k, p):\n",
    "        startingMotifs = []\n",
    "        #print(2, startingMotifs)\n",
    "        for seq in seqs:\n",
    "            startingMotifs.append(RandomMotif().motif(seq, k))\n",
    "        #print(3, startingMotifs)\n",
    "        x = ('', 0)\n",
    "        #print(4, x)\n",
    "        y = self.motifSearch(startingMotifs, seqs, k, p)\n",
    "        #print(5, y)\n",
    "        while x[0] != y[0]:\n",
    "            c = y\n",
    "            newMotifs = MotifDict().probability(seqs, profileMatrix, k)\n",
    "            #print(11, newMotifs)\n",
    "            y = self.motifSearch(newMotifs,seqs, k, p)\n",
    "            #print (12, y)\n",
    "            x = c\n",
    "        #(self.consensus, self.score) = y\n",
    "        #print(13, x, y)   \n",
    "        return y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 739,
   "id": "d978d8f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15 AAAAAAACCCAAA 18.299280225947342\n"
     ]
    }
   ],
   "source": [
    "########################################################################\n",
    "# Main\n",
    "# Here is the main program\n",
    "# This is a place to instantiate objects, read and write some data for\n",
    "# those objects and maybe implement some of the command line options.\n",
    "########################################################################\n",
    "\n",
    "def main(infile, myCommandLine=None):\n",
    "    '''\n",
    "    Get the peptide and establish the mass spectrum for the cycle.\n",
    "\n",
    "    '''\n",
    "    \n",
    "    #myCommandLine = CommandLine(myCommandLine)\n",
    "\n",
    "    c1 = CommandLine(myCommandLine)\n",
    "    i = c1.args.iterations\n",
    "    p = c1.args.psuedocount\n",
    "    k = c1.args.length\n",
    "    \n",
    "    fa = FastAreader(infile)\n",
    "    \n",
    "    seqs = []\n",
    "    for head, seq in fa.readFasta():\n",
    "        seqs.append(seq.upper())\n",
    "        #print (1, seq)\n",
    "    \n",
    "    max_value = 0\n",
    "    max_consensus = ''\n",
    "    for j in range(i):\n",
    "        outcome = MotifFinder().iterate(seqs, k, p)\n",
    "        #print(14, outcome)\n",
    "        if outcome[1] >= max_value:\n",
    "            max_value = outcome[1]\n",
    "            max_consensus = outcome[0]\n",
    "    print(15, max_consensus, max_value)\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    main('/home/nick_rose/Downloads/CRISPR_test/CONCAT.txt', ['-i 1000', '-p 1', '-k 13'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 682,
   "id": "f085aff1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nInspection\\ninspection Team\\nresults\\ncorrections\\n'"
      ]
     },
     "execution_count": 682,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Inspection\n",
    "inspection Team\n",
    "results\n",
    "corrections\n",
    "'''\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35e0f4a6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
