{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f98fb9fe",
   "metadata": {},
   "source": [
    "Rosalind_16.ipynb problem16 Nicholas Rose BME 205\n",
    "\n",
    "Peptide Encoding Problem:\n",
    "Find substrings of a genome encoding a given amino acid sequence.\n",
    "Given: A DNA string Text and an amino acid string Peptide.\n",
    "Return: All substrings of Text encoding Peptide (if any such substrings exist)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "380d3cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "class DNASeq():\n",
    "    '''\n",
    "    Creates an object (DNASeq) from a given string sequence.\n",
    "    Includes a conversion table from trinucleotides to amino acids\n",
    "    '''\n",
    "    \n",
    "    table = {'UUU':'F', 'UUC':'F', 'UUA':'L', 'UUG':'L', 'CUU':'L', 'CUC':'L', 'CUA':'L',\n",
    "            'CUG':'L', 'AUU':'I', 'AUC':'I', 'AUA':'I', 'AUG':'M', 'GUU':'V', 'GUC':'V',\n",
    "            'GUA':'V', 'GUG':'V', 'UCU':'S', 'UCC':'S', 'UCA':'S', 'UCG':'S', 'CCU':'P',\n",
    "            'CCC':'P', 'CCA':'P', 'CCG':'P', 'ACU':'T', 'ACC':'T', 'ACA':'T', 'ACG':'T',\n",
    "            'GCU':'A', 'GCC':'A', 'GCA':'A', 'GCG':'A', 'UAU':'Y', 'UAC':'Y', 'UAA':'_',\n",
    "            'UAG':'_', 'CAU':'H', 'CAC':'H', 'CAA':'Q', 'CAG':'Q', 'AAU':'N', 'AAC':'N', \n",
    "            'AAA':'K', 'AAG':'K', 'GAU':'D', 'GAC':'D', 'GAA':'E', 'GAG':'E', 'UGU':'C',\n",
    "            'UGC':'C', 'UGA':'_', 'UGC':'W', 'CGU':'R', 'CGC':'R', 'CGA':'R', 'CGG':'R', \n",
    "            'AGU':'S', 'AGC':'S', 'AGA':'R', 'AGG':'R', 'GGU':'G', 'GGC':'G', 'GGA':'G', \n",
    "            'GGG':'G'}\n",
    "    \n",
    "    def __init__ (self, DNASeq):\n",
    "        self.DNASeq = DNASeq\n",
    "        \n",
    "    def reverseCompliment(self, seq):\n",
    "        '''\n",
    "        Returns the reverse compliment of a given DNA sequence\n",
    "        '''\n",
    "        \n",
    "        compliment = ''\n",
    "        for nucleotide in seq:\n",
    "            if nucleotide == 'A':\n",
    "                compliment += 'T'\n",
    "            if nucleotide == 'T':\n",
    "                compliment += 'A'\n",
    "            if nucleotide == 'C':\n",
    "                compliment += 'G'\n",
    "            if nucleotide == 'G':\n",
    "                compliment += 'C'\n",
    "                \n",
    "        return compliment[::-1]\n",
    "        \n",
    "        \n",
    "    def transcribe(self, seq):\n",
    "        '''\n",
    "        Returns the RNA transcription of a given DNA sequence\n",
    "        '''\n",
    "        \n",
    "        RNA = ''\n",
    "        for nucleotide in seq:\n",
    "            RNA += nucleotide.replace('T', 'U')\n",
    "        return RNA\n",
    "    \n",
    "            \n",
    "    def translate(self, seq):\n",
    "        '''\n",
    "        Returns the Peptide translation of a given RNA sequence\n",
    "        '''\n",
    "        \n",
    "        peptide = ''\n",
    "        for i in range(0, len(seq), 3):\n",
    "            if seq[i:i+3] in self.table:\n",
    "                peptide += self.table[seq[i:i+3]]\n",
    "        return peptide\n",
    "                \n",
    "                \n",
    "    def findPeptide(self, seq, compliment, peptide):\n",
    "        '''\n",
    "        Method to identify DNA substrings that encode a given peptide.\n",
    "        Method takes a peptide and a DNA sequence, as well as Boolean\n",
    "        of if the DNA sequence is a reverse compliment.\n",
    "        If False, then the substrings are returned.\n",
    "        If True, then reverse compliments of the substrings are returned.\n",
    "        '''\n",
    "        \n",
    "        seqs = []\n",
    "        lenWindow = len(peptide) * 3\n",
    "        for i in range(len(seq[:-lenWindow])):\n",
    "            window = seq[i:i+lenWindow]\n",
    "            RNA = self.transcribe(window)\n",
    "            translation = self.translate(RNA)\n",
    "            \n",
    "            if translation == peptide:\n",
    "                if compliment == False:\n",
    "                    seqs.append(window)\n",
    "                if compliment == True:\n",
    "                    seqs.append(self.reverseCompliment(window))\n",
    "        \n",
    "        return seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "1b41a18f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(infile):\n",
    "    '''\n",
    "    The main method. This method takes file containing a row with\n",
    "    a peptide, and row with a DNA sequence string.\n",
    "    Prints output into a file named 'rosalind_16.txt.out', \n",
    "    containing rows of returned substrings.\n",
    "    '''\n",
    "    \n",
    "    with open(infile) as f:\n",
    "        DNA = f.readline().strip()\n",
    "        Peptide = f.readline().strip()\n",
    "    \n",
    "    seqObject = DNASeq(DNA)\n",
    "    \n",
    "    complimentDNA = seqObject.reverseCompliment(DNA)\n",
    "    \n",
    "    substrings = seqObject.findPeptide(DNA, False, Peptide)\n",
    "    substrings = substrings + seqObject.findPeptide(complimentDNA, True, Peptide)\n",
    "    substrings.sort\n",
    "    \n",
    "    with open('rosalind_16.txt.out', 'w') as out:\n",
    "            for i in substrings:\n",
    "                print(i, file=out)\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    main('/home/nick_rose/Downloads/rosalind_ba4b.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40eb8b73",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
